xgcm/sgrid.py

```diff
@@ -0,0 +1,230 @@
+        # Check for a vertical dimension
+        if "vertical_dimensions" in ds[sgrid_grid_name].attrs:
+            axes.update(["Z"])
+    elif ndims == 3:
+        axes.update(["X", "Y", "Z"])
+    else:
+        raise ValueError(
+            f"SGRID dimensions {ndims} in variable {sgrid_grid_name} is > 3."
+        )
+    return axes
+
+
+def get_axis_positions_and_coords(ds, axis_name):
+    """SGRID version of related function in `comodo.py`.
+    Returns an ordered dictionary of position->coordinate dimension entries
+    associated with a given axis, which is then stored in the `Axis.coords`
+    instance variable.
+    Populate with keywords 'center', 'inner', 'outer', 'left', 'right'
+    corresponding to the name of the dimensions on this axis.
+    For SGRID this can be done based on whether it is node or face and looking
+    at the padding attribute.
+
+    1) Node dimensions are guaranteed by basic SGRID so extract these first
+    2) Face or Volume dimensions are guaranteed and are associated with the cell
+       coordinates (center)
+        a) Check grid dimension and extract face/vol appropriately
+        b) Syntax match face/vol dimensions to node dimension
+        c) Record Shift and assign to nodes as appropriate
+    3) Facei and edgei are optional and depend on 2D vs 3D.
+       They add no additional information about dimensions so ignored for now.
+
+    SGRID cell padding types can be converted to xgcm node positions as follows:
+      - padding low  refers to 'right'
+      - padding high refers to 'left'
+      - padding both refers to 'inner'
+      - padding none refers to 'outer'
+
+    Parameters
+    ----------
+    ds : xarray.dataset
+
+    Returns
+    ----------
+    axis_name : str
+        The name of the axis to find (e.g. 'X')
+
+    References
+    ----------
+    https://sgrid.github.io/sgrid/
+    """
+
+    # Dictionary mapping SGRID padding types to xgcm node positions
+    pad2pos = {
+        "high": "left",
+        "low": "right",
+        "both": "inner",
+        "none": "outer",
+    }
+
+    # Extract the name of the SGRID grid variable and number of dimensions
+    sgrid_grid_name = get_sgrid_grid(ds)
+    sgrid_grid_dim = ds[sgrid_grid_name].attrs["topology_dimension"]
+
+    # Generate an empty dictionary to store coordinates associated with the axis
+    axis_coords = OrderedDict()
+
+    # Select the index to use corresponding to X, Y, or Z axis
+    if axis_name == "X":
+        i_select = 0
+    elif axis_name == "Y":
+        i_select = 1
+    elif axis_name == "Z":
+        i_select = 2
+    else:
+        raise ValueError(
+            f"Axis name '{axis_name}' not recognised as one of the default SGRID values 'X', 'Y', 'Z'."
+        )
+
+    # Due to the way vertical coordinates are represented with a 2D dataset
+    # using SGRID catch this case and handle before general cases:
+    if (axis_name == "Z") & ("vertical_dimensions" in ds[sgrid_grid_name].attrs):
```

Comment:
I wrote a regex based parser in https://github.com/xarray-contrib/cf-xarray/pull/421
Right now it just returns a set of dimension names, but it is capturing the padding info, just not doing anything with it.

Examples
```python
import cf_xarray as cfxr
from cf_xarray.datasets import roms_sgrid

cfxr.sgrid.parse_axes(roms_sgrid)
```
```python
{'X': {'xi_psi', 'xi_rho', 'xi_u', 'xi_v'},
 'Y': {'eta_psi', 'eta_rho', 'eta_u', 'eta_v'},
 'Z': {'s_rho', 's_w'}}
```

We could easily have `ds.cf.sgrid` return a custom object that has all the info you need here (and even more!)

Reply:
Thanks for linking this.
From what I can tell this is tailored specifically to the ROMS use of SGRID, rather than the generic case, is that correct?

Reply:
It should (in theory) work with the others because the regex is general : `re.compile("\\s?(.*?):\\s*(.*?)\\s+(?:\\(padding:(.+?)\\))?")`

Haven't tested it yet though. I'd like to add it to cf-xarray anyway because it gives us a way of identifying dimensions with axes names (X, Y, Z) when there is no metadata associated with the dimension itself.

and since it'll be in cf-xarray anyway, xgcm might as well reuse it.

Reply:
Ah, yes, gotcha.
The regex is nice.
Can add a todo to update method along with locating grid_topology identification once upstream changed are available.

Reply:
Might be nice to capture this exchange and actionable items in an issue.